WIP: Added annotations parsing and doc generation

StupidScience opened the pull request:
Fixes #3501

Hi, @ihcsim @grampelberg 
Please let me know if I am on the right way.
Is it something you expected?

StupidScience commented:
Here is corresponding PR for website - https://github.com/linkerd/website/pull/560

grampelberg commented:
Awesome, thank you!

StupidScience commented:
@grampelberg  atm it expected that it is running in the folder with downloaded repo.  Is it ok? I can try to add some flag to define path to repo. Also atm it parse all constants that have `Annotation` suffix, so it take not only proxy annotations but all of them. Also if make condition to parse constants with `Proxy` prefix it will parse some not expected annotations (e.g. `linkerd.io/inject`, `linkerd.io/version`). How do you think it should be handled?
Also if this approach is ok for you I will work further: refactor a bit and will try to add some tests.

grampelberg commented:
@ihcsim it feels a little funky to be parsing godocs to get all this (and while it isn't awful having the code checked out, feels like an extra step). Does it make sense to change `labels.go` over to being a struct that can just be YAML marshal'd into the correct spot?

adleong commented:
I will second @grampelberg's suggestion to changing `labels.go` and retrieving the descriptions programmatically.  One option would be to change the annotations from just being a string constant to being a struct containing the name and the description.  The downside to this would be we would need to update all references to the annotations.

A less invasive way would be to just create a `map[string]string` in `labels.go` which maps the annotation name to a description.

StupidScience commented:
So option with `map[string]string` looks simple: we create some public function that return it coz you can't use `map[string]string` as const. Right?
But I can't figure out how structs will help with this task. It also will require kinda function that will return array of structs or something like this.
May be i missed something, what is suggested way to get list of these structs from package?

Anyway looks like both ways lead to keeping some list/map of annotations updated (though in the same package).

adleong commented:
>Anyway looks like both ways lead to keeping some list/map of annotations updated (though in the same package).

Yeah, I think that's right.  Some kind of function which returns a list of structs, or some kind of map.  You're on the right track :+1:

StupidScience commented:
@grampelberg @adleong I've reworked it to map. Can you take a look?
Now it looks like I just moved this functionality from site generation with Hugo to `linkerd doc` cmd.

adleong commented:
Nice work @StupidScience!  However, I'm not sure if we want to assume that flag names and annotation names will match, even though they do in most cases.  I'd personally be more comfortable if the annotation descriptions were standalone, even though that will involve some duplication with the flag documentation.  @ihcsim what do you think about that?

StupidScience commented:
@ihcsim @adleong  I am still waiting for your feedback.

ihcsim commented on cli/cmd/doc.go:
```diff
@@ -93,3 +111,42 @@ func generateDocs(cmd *cobra.Command) ([]cmdDoc, error) {
+func generateAnnotationsDocs(annotationsMap map[string]string, cmdList []cmdDoc) []annotationDoc {
+	var annotationsList []annotationDoc
+
+	for _, cmd := range cmdList {
```
There are times when we will add a proxy annotation w/o the corresponding CLI flag. Distributed tracing is an example, where the CLI flag is added later. (That is also why you had to explicitly add the tracing-related descriptions to `GetAnnotationsDocs()` and have a second `for` loop in line 142 below).

I think for simplicity sake,

1. let's move `GetAnnotationsDocs()` to this file since it's not used anywhere else
1. make it an unexported struct, and
1. as @adleong suggested, reuse the CLI commands descriptions in the new struct (I don't mind the duplication being confined in an unexported struct)

Ideally, there should be some sort of struct/map (in a new package under `pkg`) that represents the relationship of the annotation, the CLI option and the description. But since the `proxyConfigOptions` in `cli/cmd/root.go` is unexported and used in many places, modifying it will increase the change scope of this PR.

ihcsim commented on pkg/k8s/labels.go:
```diff
@@ -334,3 +334,33 @@ func IsTapDisabled(pod *corev1.Pod) bool {
+
+// GetAnnotationsDocs return map with annotations and its descriptions
+// description can be omitted so it will be taken from corresponding cli flag
+func GetAnnotationsDocs() map[string]string {
```
Since all the keys in the map are exported and this is only used in `cli/cmd/doc.go`, we can just move this to that file, so that it's close to where it's used.

ihcsim commented:
> Yeah, I think this is perfectly fine. We can just mention that the linkerd doc command must be run in the linkerd2 project root, in the relevant section in BUILD.md.

I just notice that this is no longer relevant since we are no longer using the godoc approach.

StupidScience commented on cli/cmd/doc.go:
```diff
@@ -7,40 +7,58 @@ import (
-	Shorthand    string `yaml:",omitempty"`
-	DefaultValue string `yaml:"default_value,omitempty"`
-	Usage        string `yaml:",omitempty"`
+	Shorthand    string
```
Looks like it is kinda legacy from other `yaml` package and these tags are not used by `"sigs.k8s.io/yaml"`

adleong replied:
ah, it's unintentional that these options are not being applied.  it looks like we currently have a bunch of `SeeAlso: null` in the output, for example.  we should look into how to omit these fields in "sigs.k8s.io/yaml".

StupidScience replied:
I've investigated a bit and `"sigs.k8s.io/yaml"` accept json-tags.
So here should be `json:",omitempty"` and so on to make it work. As for me it'll look confusingly.

ihcsim replied:
> I've investigated a bit and "sigs.k8s.io/yaml" accept json-tags.

Yeah, my jaw dropped when I first found out about it, because it first converts everything to JSON and then into YAML.

StupidScience commented:
@ihcsim I've reworked it according to your comments. Can you please take a look one more time?
